processing/src/main/java/org/apache/druid/segment/serde/ComplexMetrics.java

```diff
@@ -38,12 +38,18 @@ public static ComplexMetricSerde getSerdeForType(String type)
     return complexSerializers.get(type);
   }
 
-  public static void registerSerde(String type, Supplier<ComplexMetricSerde> serdeSupplier)
+  public static <T extends ComplexMetricSerde> void registerSerde(String type, Class<T> clazz, Supplier<T> serdeSupplier)
   {
-    if (ComplexMetrics.getSerdeForType(type) == null) {
-      if (complexSerializers.containsKey(type)) {
-        throw new ISE("Serializer for type[%s] already exists.", type);
+    if (complexSerializers.containsKey(type)) {
+      if (!complexSerializers.get(type).getClass().getCanonicalName().equals(clazz.getCanonicalName())) {
```

Comment:
can't we check  the serde classname from serdeSupplier.get().getClass().getName() ?
in fact, I am not sure if even taking a Supplier is necessary all of this could possibly be done with original method signature which was `registerSerde(String type, ComplexMetricSerde serde)` by using `serde.getClass().getName()` or am I missing something ?

Reply:
Good point, I'm not actually certain why it switched to a supplier, I'll play around with this soon to see if I can simplify

Reply:
Switched back to the original signature. Any reason to use `getName` instead of `getCanonicalName`?

Reply:
it wouldn't matter in most cases except if someone's custom extension has an anonymous class impl like `new ComplexMetricSerde() { .... impl ... }` . In that case `getCanonicalName()` returns null while `getName()` returns something like "x.y.A$1" where "x.y.A" is the class name where that anonymous class was defined.

Reply:
Ah, good point, will change :+1:

Reply:
Such non-obvious considerations should have been reflected in code comments along with the code, not just remain in PR discussions. Otherwise, future developers might use `getCanonicalName()` without noticing a problem.

Reply:
>Such non-obvious considerations should have been reflected in code comments along with the code, not just remain in PR discussions. Otherwise, future developers might use getCanonicalName() without noticing a problem.

Ah, I agree, I'll open a PR adding some comments and linking to this discussion sometime soon.

Reply:
I think this might be more general than this context and better served as a static analysis tool check to never use `getCanonicalName()` and use `getName()` instead of some comment here.

Reply:
`getCanonicalName()` is preferred during `toString()` and logging operations: see https://stackoverflow.com/a/15203417.

Reply:
hmmm...thanks for that link, read through the answer there, I see that in case of arrays `getCanonicalName()` returns something more recognizable however it returning null for anonymous class is a turnoff for me as that leads to NPE and causes bugs.
I don't think that output of `getName()` is totally unrecognizable for toString().  So I still think it is better to make it a rule and use `getName()` even in toString().
Also, I am not sure but I think those tools would allow occasional exceptions to the rule e.g. if there is a critical situation where it is absolute necessary to use `getCanonicalName()` then that could happen with some comment for the tool to ignore that occurrence .

Reply:
for funzies :) , I also checked how much of `getCanonicalName()` are we using in our code...

```
$ grep -r 'getCanonicalName()' ./
.//src/main/java/org/apache/druid/guice/GuiceInjectableValues.java:        valueId.getClass().getCanonicalName(),
.//src/main/java/org/apache/druid/java/util/common/lifecycle/Lifecycle.java:              .getCanonicalName())) {
.//src/main/java/org/apache/druid/java/util/common/lifecycle/Lifecycle.java:              .getCanonicalName())) {
.//src/main/java/org/apache/druid/metadata/DefaultPasswordProvider.java:    return this.getClass().getCanonicalName();
```

it will make no real difference if they were all replaced by `getName()` .

Reply:
`getName` (or `getTypeName` i guess) seem more useful/informative to me as well for the same reason about nulls. It is a bit less friendly, but an error is probably never really friendly, so it least this way we can be sure it is informative in helping isolate the issue that caused the error. 

I found a few more usages of `getCanonicalName` searching with intellij, but they are all similar usages to the ones you listed and could probably also safely be adapted. I don't feel strongly either way about a rule to force it, but am willing to add it if that is best.

Reply:
I accidentally did my search in "core/" so grep missed  .. it wasn't grep's but my fault :)

so I ran grep again and I see that there are about 50 occurrences of `getCanonicalName()` in the code ... took a cursory look at them and don't see  any harm converting them to `getName()`

Reply:
Looked a bit closer, I don't think we can _fully_ remove usage of `getCanonicalName`, extension config list modules are done with canonical name in `Initialization.java`.

```
    private void tryAdd(T serviceImpl, String extensionType)
    {
      final String serviceImplName = serviceImpl.getClass().getCanonicalName();
      if (serviceImplName == null) {
        log.warn(
            "Implementation [%s] was ignored because it doesn't have a canonical name, "
            + "is it a local or anonymous class?",
            serviceImpl.getClass().getName()
        );
...
```

Should I add a rule to forbid it anyway and just ignore this legitimate usage?

Reply:
I think yes

Reply:
@clintropolis 
I think even that can be replaced, that string is only being used for deduplication which can be done from output of `getName()` as well. Did you get any error when changing that occurrence to `getName()` ?

Reply:
>I think even that can be replaced, that string is only being used for deduplication which can be done from output of getName() as well. Did you get any error when changing that occurrence to getName() ?

I haven't tried yet, will have a look :+1:

Reply:
Tests passed, you're right, can ban it entirely, will open a PR soon.

Reply:
Raised #8086